### PR TITLE
Re-organize render (part 1)

### PR DESCRIPTION
**Pull request by frankmcsherry**
The first part of attempts to smash up `render/mod.rs` into smaller parts. These are the easier parts that were already in their own methods.

One relevant change is that the methods `render_foo` should be called only after ensuring that their inputs are rendered. This is because the methods can only make that recursive call back in to the `Context` if they have the most general type parameters, which is more restrictives than we actually require. The change allows for implementations that are more clearly `Timestamp`-independent, and also a little less shuffling around of things like scopes and worker ids (important for some operators, not for others).

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/2916)

### CODE REVIEW

File: src/dataflow/render/mod.rs [hunk 1]
```diff
@@ -107,10 +107,10 @@ use std::rc::Weak;
 use differential_dataflow::hashable::Hashable;
 use differential_dataflow::lattice::Lattice;
 use differential_dataflow::operators::arrange::arrangement::Arrange;
-use differential_dataflow::operators::arrange::arrangement::ArrangeByKey;
+// use differential_dataflow::operators::arrange::arrangement::ArrangeByKey;
```
Comment (umanwizard) [inline src/dataflow/render/mod.rs]: stray?

File: src/dataflow/render/mod.rs [hunk 2]
```diff
@@ -107,10 +107,10 @@ use std::rc::Weak;
...
 use differential_dataflow::operators::arrange::upsert::arrange_from_upsert;
-use differential_dataflow::operators::join::JoinCore;
-use differential_dataflow::trace::implementations::ord::{OrdKeySpine, OrdValSpine};
+// use differential_dataflow::operators::join::JoinCore;
+// use differential_dataflow::trace::implementations::ord::{OrdKeySpine, OrdValSpine};
```
Comment (umanwizard) [inline src/dataflow/render/mod.rs]: Also here

File: src/dataflow/render/mod.rs [hunk 3]
```diff
@@ -983,528 +1006,16 @@ where
...
-                                    .chain(next_vals.iter().map(|i| &next_datums[*i])),
-                            ))
-                        });
-                        errs = errs.concat(&es.as_collection(|k, _v| k.clone()));
-                    }
-                    None => {
-                        panic!("Arrangement alarmingly absent!");
+                RelationExpr::ArrangeBy { input, keys } => {
+                    if keys
+                        .iter()
```
Comment (umanwizard) [inline src/dataflow/render/mod.rs:1008]: Earlier we were calling `ensure_rendered` on the input of the `keys` vector was empty; why is this no longer necessary?

File: src/dataflow/render/reduce.rs [hunk 4]
```diff
@@ -17,26 +17,29 @@ use differential_dataflow::operators::{Reduce, Threshold};
...
+    T: Timestamp + Lattice,
 {
     /// Renders a `RelationExpr::Reduce` using various non-obvious techniques to
     /// minimize worst-case incremental update times and memory footprint.
     pub fn render_reduce(
         &mut self,
         relation_expr: &RelationExpr,
         scope: &mut G,
-        worker_index: usize,
+        // worker_index: usize,
```
Comment (umanwizard) [inline src/dataflow/render/reduce.rs]: stray

File: src/dataflow/render/threshold.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+impl<G, T> Context<G, RelationExpr, Row, T>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Refines<T>,
+    T: Timestamp + Lattice,
+{
+    pub fn render_threshold(
+        &mut self,
+        relation_expr: &RelationExpr,
+        // scope: &mut G,
```
Comment (umanwizard) [inline src/dataflow/render/threshold.rs]: stray

File: src/dataflow/render/threshold.rs [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        // worker_index: usize,
+    ) {
+        if let RelationExpr::Threshold { input } = relation_expr {
+            // TODO: re-use and publish arrangement here.
+            let arity = input.arity();
+            let keys = (0..arity).collect::<Vec<_>>();
+
+            // TODO: easier idioms for detecting, re-using, and stashing.
+            if self.arrangement_columns(&input, &keys[..]).is_none() {
+                // self.ensure_rendered(input, scope, worker_index);
```
Comment (umanwizard) [inline src/dataflow/render/threshold.rs:36]: stray

File: src/dataflow/render/top_k.rs [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+            input,
+            group_key,
+            order_key,
+            limit,
+            offset,
+        } = relation_expr
+        {
+            use differential_dataflow::operators::reduce::Reduce;
+
+            // self.ensure_rendered(input, scope, worker_index);
```
Comment (umanwizard) [inline src/dataflow/render/top_k.rs:39]: stray

File: src/dataflow/render/arrange_by.rs [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+impl<G, T> Context<G, RelationExpr, Row, T>
+where
+    G: Scope,
+    G::Timestamp: Lattice + Refines<T>,
+    T: Timestamp + Lattice,
+{
+    pub fn render_arrangeby(
+        &mut self,
+        relation_expr: &RelationExpr,
+        // scope: &mut G,
```
Comment (umanwizard) [inline src/dataflow/render/arrange_by.rs]: stray

File: src/dataflow/render/mod.rs (hunk 3 above)
Comment (frankmcsherry) [reply src/dataflow/render/mod.rs:1008]: Good question! Probably a bug. Tbh, it's not clear what an arrangement with no keys means (the list contains the requested arrangements), though that should probably be documented one way or the other. Will make sure it maintains the same behavior for now.

File: src/dataflow/render/mod.rs [hunk 9]
```diff
@@ -983,528 +1003,17 @@ where
...
-                            .iter()
-                            .find(|equivs| equivs.contains(expr))
-                            .expect("Expression in join plan is not in an equivalence relation");
-
-                        // We expect to find exactly one bound expression, as
-                        // multiple bound expressions should result in a filter
-                        // and be removed once they have.
-                        let mut bound_expr = equivalence
+                RelationExpr::ArrangeBy { input, keys } => {
+                    if keys.is_empty()
```
Comment (umanwizard) [inline src/dataflow/render/mod.rs:1007]: Can you add a comment saying we're not really sure this is necessary and just maintaining the old behavior

Comment (frankmcsherry) [reply src/dataflow/render/mod.rs:1007]: It is currently required based on the docs for `RelationExpr::ArrangeBy` which says that it is a no-op (and by my reading, should ensure that `input` is available independent of its arguments). I'll add an explanation.

**Pull request merged**